src/gf_ring.h

```diff
@@ -148,12 +148,24 @@ class RingModN {
     std::vector<int> exponents;
     std::vector<T> all_primes_factors;
     std::vector<T> proper_divisors;
+    // random number engine with a seed as a random number
+    mutable std::uniform_int_distribution<uint64_t> distribution;
+    // random number engine for the __uint128_t case
+    mutable std::uniform_int_distribution<uint64_t> distribution_leading;
```

Comment:
Why do you need this one? You could simply make two calls to `distribution` for the 128-bit case, no?

Reply:
The two distributions are used for two different purposes.

- If `q <= std::numeric_limits<uint64_t>::max()`, only `distribution` is enough.

- Otherwise,  we use `distribution` and `distribution_leading` to generate random number `x = x1 + x2` where

    - `x1 in [1, std::numeric_limits<uint64_t>::max()` is generated by `distribution`
    - `x2 in [0, q-std::numeric_limits<uint64_t>::max()`) is generated by`distribution_leading`
 
Oups, I saw a bug to fix :)

Reply:
Ha, indeed, then it makes sense to have two distribution.